test/dict.jl

```diff
@@ -388,6 +388,11 @@ d = Dict('a'=>1, 'b'=>1, 'c'=> 3)
 @test [d[k] for k in keys(d)] == [d[k] for k in eachindex(d)] ==
       [v for (k, v) in d] == [d[x[1]] for (i, x) in enumerate(d)]
 
+# generators, similar
+d = Dict(:a=>"a")
+@test @inferred(map(identity, d)) == d
+@test @inferred(map(p->p.first=>p.second[1], d)) == Dict(:a=>'a')
+@test @inferred(map(p->p.second, d)) == Any["a"]
```

Comment:
Yes. I would be against sometimes mapping a Dict to a Dict, and sometimes to an array. Also I believe the `collect` machinery assumes the container type doesn't depend on the element type --- as it runs, it might pick a new element type but keeps the container type.